engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/SetWarningsNode.java

```diff
@@ -0,0 +1,118 @@
+package org.enso.interpreter.runtime.warning;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Cached.Shared;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.enso.interpreter.dsl.AcceptsWarning;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.hash.EnsoHashMap;
+import org.enso.interpreter.runtime.data.hash.HashMapInsertNode;
+
+@BuiltinMethod(
+    type = "Warning",
+    name = "set_array",
+    description = "Sets all the warnings associated with the value.",
+    autoRegister = false)
+@GenerateUncached
+public abstract class SetWarningsNode extends Node {
+
+  public static SetWarningsNode build() {
+    return SetWarningsNodeGen.create();
+  }
+
+  public abstract Object execute(VirtualFrame frame, @AcceptsWarning Object value, Object warnings);
+
+  @Specialization(guards = "isEmpty(warnings, interop)")
+  Object doEmpty(
+      VirtualFrame frame,
+      Object object,
+      Object warnings,
+      @Shared @CachedLibrary(limit = "3") InteropLibrary interop,
+      @Shared @Cached ConditionProfile isWithWarnsProfile) {
+    if (isWithWarnsProfile.profile(object instanceof WithWarnings)) {
+      return ((WithWarnings) object).value;
+    } else {
+      return object;
+    }
+  }
+
+  @Specialization(guards = "!isEmpty(warnings, interop)")
+  Object doSetArray(
+      VirtualFrame frame,
+      Object object,
+      Warning[] warnings,
+      @Shared @CachedLibrary(limit = "3") InteropLibrary interop,
+      @Shared @Cached HashMapInsertNode mapInsertNode,
+      @Shared @Cached ConditionProfile isWithWarnsProfile) {
```

Comment:
There is [InlinedConditionProfile](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/profiles/InlinedConditionProfile.html) now in the Truffle API. It may be more lightweight and the rewrite seems straightforward.